cibersort/cibersort.xml

```diff
@@ -0,0 +1,170 @@
+  levelColor = paste0(level,"Color")
+  if ("$mapping.provided"=="FALSE" || toupper("$mapping.map_colors")==F 
+      || is.na(map) || is.na(level) || !(levelColor %in% colnames(map))
+      || is.na(map[,levelColor])) {
+    colorCount = length(cellTypes)
+    getPalette = colorRampPalette(brewer.pal(9, "Set1"))
+    colors = rev(getPalette(colorCount))
+  } else {
+    colors = setNames(as.character(map[,levelColor]), map[,level])
+  }
+  return(colors)
+}
+
+# Define the function that extracts an array of valid column titles for the bulk
+# file
+convertBulkColNames = function(samples) {
+    asEntered = unlist(strsplit(samples, ","))
+    subDots = lapply(asEntered,FUN=function(x){return(gsub('-','\\.',x))})
+    addX = lapply(subDots,FUN=function(x){return(paste0('X',x))})
+    return(unlist(addX))
+}
+
+# Set up required data
+signature = read.csv("$signature_file", header=TRUE, row.names=1, sep="\t")
+bulk = as.matrix(read.csv("$expression_file", header=TRUE, row.names=1, sep="\t"))
+samples = convertBulkColNames("$sample_ids")
+##bulk = as.matrix(select(bulk, unlist(samples)))
+bulk = as.matrix(bulk)
+
+res = CIBERSORT(signature, bulk,QN=TRUE, absolute=FALSE, abs_method="sig.score")
+write.table(res, "$output_tsv", quote=FALSE, sep="\t")
+
+cols = ncol(signature)
+cells = res[,1:cols] *100
+cells = cells[samples,,drop=FALSE]
+data = melt(cells)
+colnames(data) = c("Sample", "CellType", "EstimatedPercent")
+
+#if $mapping.provided == "FALSE"
+colors = getColors(cellTypes=unique(data$CellType))
+graph(data, "CellType", colors=colors)
+#else
+groups = read.csv("$mapping.map_file", header=TRUE, sep="\t", row.names=NULL,stringsAsFactors=TRUE)
+merged = merge(data, groups, by="CellType")  
+levels = unlist(strsplit("$mapping.resolutions",","))#('CellType', 'Full', 'Condensed')
+for (level in levels) {
+  # Aggregate cell types to appropriate granularity
+  cellTypes = unique(groups[,level])
+  formula = as.formula(paste0(c("EstimatedPercent ~ Sample + ",level)))
+  summedData = aggregate(formula, merged, FUN=sum) 
+  
+  # Sort the cell types for the purpose of
+  #colnames(summedData) = c("CellType", "Sample", "EstimatedPercent")
+  summedData[,level] = as.character(summedData[,level])
+  summedData[,level] = factor(summedData[,level], levels=cellTypes)
+  summedData[order(summedData[,level]),] 
+
+  # Determine colors and graph
+  ##summedData = summedData[summedData[,"Sample"]=="$sample_id"]
+  colors = getColors(cellTypes, map=groups, level=level)
+  graph(summedData, level, colors)
+}
+#end if
+        </configfile> 
+    </configfiles>
+
+    <inputs>
+        <param name="signature_file" type="data" format="tsv" label="Gene Signature File" />
+        <param name="expression_file" type="data" format="tsv" label="Cohort Gene Expression"
+               help="Hugo-named gene expression file" />
+        <param name="sample_ids" type="text" optional="false" label="Sample ID" />
+        <conditional name="mapping">
```

Comment:
I think we discussed changing/rephrasing the conditional here. We want to ask the user if they would like the default cibersort cell types or if they would like to use custom cell types. For the latter case, they need to provide a mapping file of cell types and colors as well as the cell type granularity.